nixos/modules/services/web-apps/keycloak.nix

```diff
@@ -0,0 +1,441 @@
+
+        systemd.services.keycloakDatabaseInit = lib.mkIf databaseActuallyCreateLocally {
+          after = [ "postgresql.service" ];
+          before = [ "keycloak.service" ];
+          bindsTo = [ "postgresql.service" ];
+          serviceConfig = {
+            Type = "oneshot";
+            RemainAfterExit = true;
+            User = "postgres";
+            Group = "postgres";
+          };
+          script = ''
+            set -eu
+
+            if ! [[ -e ${cfg.databasePasswordFile} ]]; then
+                echo "Database password file ${cfg.databasePasswordFile} not found! Aborting." >&2
+                exit 1
+            fi
+
+            PSQL=${config.services.postgresql.package}/bin/psql
+
+            db_password="$(<'${cfg.databasePasswordFile}')"
+
+            $PSQL -tAc "SELECT 1 FROM pg_roles WHERE rolname='${cfg.databaseUsername}'" | grep -q 1 || $PSQL -tAc "CREATE ROLE ${cfg.databaseUsername} WITH LOGIN PASSWORD '$db_password' CREATEDB"
+
+            $PSQL -tAc "SELECT 1 FROM pg_database WHERE datname = 'keycloak'" | grep -q 1 || $PSQL -tAc 'CREATE DATABASE "keycloak" OWNER "${cfg.databaseUsername}"'
+            current_owner=$($PSQL -tAc "SELECT pg_catalog.pg_get_userbyid(datdba) FROM pg_catalog.pg_database WHERE datname = 'keycloak'")
+            if [[ "$current_owner" != "${cfg.databaseUsername}" ]]; then
+                $PSQL -tAc 'ALTER DATABASE "keycloak" OWNER TO "${cfg.databaseUsername}"'
+                if [[ -e "${config.services.postgresql.dataDir}/.reassigning_keycloak" ]]; then
+                    echo "Reassigning ownership of database keycloak to user ${cfg.databaseUsername} failed on last boot. Failing..." >&2
+                    exit 1
+                fi
+                touch "${config.services.postgresql.dataDir}/.reassigning_keycloak"
+                $PSQL "keycloak" -tAc "REASSIGN OWNED BY \"$current_owner\" TO \"${cfg.databaseUsername}\""
+                rm "${config.services.postgresql.dataDir}/.reassigning_keycloak"
+            fi
+          '';
+        };
+
+        systemd.services.keycloak = {
+          after = lib.optionals databaseActuallyCreateLocally [
+            "keycloakDatabaseInit.service" "postgresql.service"
+          ];
+          bindsTo = lib.optionals databaseActuallyCreateLocally [
+            "keycloakDatabaseInit.service" "postgresql.service"
+          ];
+          wantedBy = [ "multi-user.target" ];
+          environment = {
+            JBOSS_LOG_DIR = "/var/log/keycloak";
+            JBOSS_BASE_DIR = "/var/lib/keycloak";
+            JBOSS_MODULEPATH = "${cfg.package}/modules";
+          };
+          serviceConfig = {
+            ExecStartPre = systemdPrefixScript "startPre" "+" (''
+              set -eu
+
+              if ! [[ -d /var/lib/keycloak/configuration ]]; then
+                  cp -r ${cfg.package}/standalone/configuration /var/lib/keycloak/
+              fi
+
+              if ! [[ -e ${cfg.databasePasswordFile} ]]; then
```

Comment:
As above, I would prefer declarative [ConditionPathExists](https://www.freedesktop.org/software/systemd/man/systemd.unit.html#Conditions%20and%20Asserts) over imperative `bash` statements.

Reply:
I changed this for the `keycloak` unit, but I'm not sure that it's an improvement - instead of getting an error when the file doesn't exist you now get a silent error that looks like a success.

Reply:
Reverted to checking with `bash`, since it makes runtime errors more explicit.

Reply:
@talyz unfortunate, but that makes sense to me. I'll ping some `systemd` keeners like @andi, @flokli, @arianvp, etc... just to make sure they don't have any feedback I've missed on `ConditionPathExists`.

Reply:
I found a solution - for every `ConditionX` there's also an `AssertX` which fails the unit instead of skipping it. It still doesn't give great error output: it only logs `keycloak.service: Starting requested but asserts failed.`, but if you check the status of the unit you can see the failed condition, so still much better than just skipping it.

Reply:
Why do we need to assert in the first place?  Doesn't keycloak already crash with a happy error message if these files are missing?

Reply:
Well, the initial reason for adding them were to get a better error message than
```
/nix/store/ad05hhpmg2mhn6jqpdgpn1c8xllqdgmk-unit-script-keycloakDatabaseInit-start/bin/keycloakDatabaseInit-start: line 6: /run/keys/db_password: No such file or directory
```
but I suppose that might actually be less cryptic than the assert message.

Reply:
I removed the asserts. If any of the files are configured incorrectly keycloak won't be started and the error message will be pretty clear.